Add initial version of the language reference

tsandall opened the pull request:
See commit message for details.

tsandall commented:
Also, I have referenced an example data set ("Acme Corp") which doesn't exist yet.

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,388 @@
+# Opalog: OPA's Query Language
+
+OPA includes a policy engine which is purpose built for reasoning about information represented in structured documents such as JSON. Data stored in in the policy engine can be queried using OPA's native query language: Opalog.
```
One of my few pet peeves in english: that vs. which.  Here's my understanding.  "that" is for part of a sentence that cannot be removed--it helps define the subject.  "which" is for part of a sentence that could be removed, and is preceded by a comma or used as a parenthetical.  Many places throughout the doc use "which" where "that" seems correct.

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,388 @@
+
+## What is Opalog?
+
+Opalog is a modernized version of [Datalog](https://en.wikipedia.org/wiki/Datalog) which is a well understood, decades old query language. Opalog extends Datalog to support structured document models such as JSON. Opalog queries are assertions on data stored in OPA. These queries can be used to define policies which enumerate violations that exist in the data stored in OPA.
```
Maybe we should say "was inspired by" instead of "is a modernized version of".

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,388 @@
+Opalog focuses on providing powerful support for referencing nested documents and ensuring that queries are correct and unambiguous.
+
+Opalog is declarative so policy authors can focus on what queries should return rather than how queries should be executed. These queries are simpler and more concise than the equivalent in an imperative language. Like other applications which support declarative query languages, OPA is able to optimize queries to improve performance, e.g, indexing, concurrent evaluation, reordering, etc.
+
```
I like this whole section

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,388 @@
+import         ::= "import" package [ "as" var ]
+policy         ::= { rule }
+rule           ::= rule-head ":-" rule-body
+rule-head      ::= rule-name [ rule-args ]
```
Can we use 'expr' for rule-head?

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,388 @@
+rule-args      ::= "[" [ key ] "]" [ "=" value ]
+key            ::= var
+value          ::= var
+rule-body      ::= [ literal { "," literal } ]
```
I'd be tempted to inline this rule-body.

Overall, the fewer concepts the better

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,388 @@
+rule-head      ::= rule-name [ rule-args ]
+rule-name      ::= var
+rule-args      ::= "[" [ key ] "]" [ "=" value ]
+key            ::= var
```
I'd imagine we'd want to allow key/value to be scalars in addition to vars.

tsandall replied:
I take this to mean having a constant in the head, e.g, `t["foo"] = v :- ...`. I don't see a problem with this on it's own however the same thing could be achieved with a variable in the head bound to a constant in the body, e.g, `t[k] = v :- k = "foo", ...`. Is there another reason to allow scalars in the head? If not, I would say let's leave it out for now.

tim-styra replied:
Happy to leave out the scalars in the head for now.

tim-styra replied:
I'd be tempted to collapse 'rule' as follows...
var [ rule-args ] :- rule-body
rule-args :- "[" [var] "]" [ "=" value ]

tim-styra replied:
Notationally, is there another symbol we can use besides [] to mean "optional"?  Isn't a question-mark standard?  And isn't an asterisk standard for repetition?  I find the use of [] and {} confusing since I'm accustomed to ?*, and since [] and {} are meaningful symbols in our language.

tsandall replied:
I will colapse the rule as you suggested. I presume "value" can also be replaced with "var".

I think "?" and "*" are more classical, however, I like this style for optional/repeated groups. I will update this section to include a link to the wikipedia page for EBNF. As a side note, Golang uses this notation.

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,388 @@
+                  | expr-infix
+expr-builtin   ::= var "(" [ term { , term } ] ")"
+expr-infix     ::= term bool-operator term
+term           ::= ref | var | const
```
We probably want dictionary and array constructors too as 'terms'

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,388 @@
+ref-arg-brack  ::= "[" ( const | var ) "]"
+ref-arg-dot    ::= "." var
+var            ::= ALPHA { ALPHA | DIGIT | "_" }
+const          ::= scalar | object | array
```
Should const include variables here?  const -> array -> term -> var

tim-styra commented on docs/LANGUAGE.md:
````diff
@@ -0,0 +1,388 @@
+# ["prod", "smoke1", "dev"]
+```
+
+We can rewrite the previous rule "r" to make use of "q". We will call the new rule "p":
````
I got a bit confused on a quick read about "previous rule".

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,388 @@
+Opalog is declarative so policy authors can focus on what queries should return rather than how queries should be executed. These queries are simpler and more concise than the equivalent in an imperative language. Like other applications which support declarative query languages, OPA is able to optimize queries to improve performance, e.g, indexing, concurrent evaluation, reordering, etc.
+
+## Opalog Grammar
+
```
It seems that this grammar is part of the introduction to the language, which would mean we should focus on making the grammar easy to understand.  10-12 lines is probably the largest anyone will try to scan through.  Maybe we could put the actual grammar at the end of the page and add a 'simplified grammar' for people to scan through.  Or maybe a short 'Overview' section here instead.

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,388 @@
+expr-infix     ::= term bool-operator term
+term           ::= ref | var | const
+bool-operator  ::= "=" | "!=" | "<" | ">" | ">=" | "<="
+ref            ::= ns { ref-arg }
```
Seems 'ns' isn't used anywhere else.  Does it make sense to replace 'ns' with 'var' here?

tsandall replied:
Yes, this makes sense. Especially since packages and import aliases are defined in terms of "var".

tsandall replied:
Yes, this can be dropped. Especially since packages and import aliases are defined in terms of var.

tim-styra commented on docs/LANGUAGE.md:
````diff
@@ -0,0 +1,388 @@
+We can generalize the above example by defining a **set rule** instead of a **boolean rule**:
+
+```opalog
+q[] = x :- 
````
I worry about starting with a set example, just because it's the most confusing syntactically.  A dictionary or array would  be better.

tim-styra commented on docs/LANGUAGE.md:
````diff
@@ -0,0 +1,388 @@
+DIGIT  ASCII characters 0-9
+```
+
+## Opalog Queries
````
I'd be tempted to first describe how Opalog lets people query base documents (without differentiating base and virtual documents).  And then say that we can also use Opalog to construct new, virtual documents out of the documents we have been given.  It so happens that the query language you have been given suffices to define those new documents.

tim-styra commented on docs/LANGUAGE.md:
````diff
@@ -0,0 +1,388 @@
+```
+
+### Terms
+
````
What are Terms?  How are they used in Opalog?

tim-styra commented on docs/LANGUAGE.md:
````diff
@@ -0,0 +1,767 @@
+```opalog
+sites :- [{"name": "prod"}, {"name": "smoke1"}, {"name": "dev"}]
+r :- sites[].name = "prod"
+```
````
I think we want this to be sites[i].name = "prod".  At least at this point, the empty braces are confusing.  And here we probably want to say that we are implicitly enumerating the elements of the sites array.

tsandall replied:
Done.

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,767 @@
+Scalar values are the simplest type of term in Opalog. Scalar values can be
+strings, numbers, booleans, or null.
+
+Rules can be defined solely in terms of scalar values. This is useful for
```
s/Rules/Documents

tsandall replied:
Done

tim-styra commented on docs/LANGUAGE.md:
````diff
@@ -0,0 +1,767 @@
+```
+
+These rules can be queried like any other:
+
````
s/rules/documents

tsandall replied:
Done

tim-styra commented on docs/LANGUAGE.md:
````diff
@@ -0,0 +1,767 @@
+pi         :- 3.14159
+allowed    :- true
+sentinel   :- null
+```
````
Just had a thought.
Instead of
pi :- 3.14159
we could write
pi = 3.14159 :- true
Then if we drop the body, which is unnecessary, we get..
pi = 3.14159

Then syntactically, every statement is a rule and a rule is of the form...
var[x] = y :- body
where body is optional, as is the [x]

tsandall replied:
+1 for this. I think it will be clearer if I address this in a subsequent PR that addresses the conceptual changes around sets and type insulation.

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,767 @@
+head and body of rules.
+
+Variables appearing in the head of a rule can be thought of as input and
+output of the rule. For example:
```
Probably worth expanding on the role of inputs and outputs with variables.

Variables appearing in the head of a rule can be thought of as both the inputs and outputs of a rule.  Unlike traditional programming languages, where a variable is either an input or an output, in Opalog a variable is simultaneously an input and an output.  If a query supplies a value for a variable, that variable is an input, and if the query does not supply a value for a variable, that variable is an output.

tsandall replied:
Done. The extra prose is good because it is explicit and then the example can help the reader build up intuition.

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,767 @@
+
+1. Non-string keys such as numbers, booleans, and null.
+2. Variable keys which are described later.
+
```
3. String keys with non alpha-numeric symbols (or whatever the grammar says is not allowed after the .).

tsandall replied:
Done.

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,767 @@
+References are always prefixed with a variable that identifes the root
+document. In the example above this is "p". The root document may be:
+
+- a local variable.
```
Perhaps...

a local variable *inside the rule*

tsandall replied:
Done.

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,767 @@
+
+Opalog allows rules to iterate sets, arrays, and objects using the same
+square bracket syntax (e.g., "sites[]" implicitly iterates all elements in "sites").
+
```
I think we want to make the variables that we are iterating over explicit.  And then treat the [] syntax as syntactic sugar.

tsandall replied:
Done.

tim-styra commented on docs/LANGUAGE.md:
````diff
@@ -0,0 +1,767 @@
+```
+apps_by_hostname["helium"]
+# "web"
+```
````
We should probably describe what happens when a single key is assigned 2 values.  Not sure what the right thing to do here is.  The top-down algorithm doesn't actually care since if you ask for all the y such that obj["abc"] = y, it will happily return multiple values.  Or if you ask if obj["abc"]="def" it will return true, and if you ask if obj["abc"] = "ghi" it will return true.  The only time multiple values for a single key will show up is when we need to construct a single dictionary of the key/value pairs to hand off to something else.

tsandall replied:
We definitely have to describe the behaviour in this case, however, I think the desired behaviour requires a bit of thought and so I will defer to a subsequent PR.

tim-styra commented on docs/LANGUAGE.md:
```diff
@@ -0,0 +1,767 @@
+
+> OPA will reorder expressions to ensure that negated expressions are evaluated after other non-negated expressions with the same variables. OPA will reject rules containing negated expressions that do not meet the safety criteria described above.
+
+The simplest use of negation involves involves only scalar values or or variables and is equivalent to complementing the operator:
```
involves involves

tsandall replied:
Done.

tim-styra commented:
Overall, this is great!  The structure is clear, and I like how there's the brief overview at the beginning followed by a deep-dive into each of the important aspects of the language.  And it's all example driven.  Really happy overall.

tsandall commented:
I have addressed all of the comments above.

I will come back with a PR (hopefully later today) that updates the documentation with the conceptual changes around sets. This will also update the various places in this doc that use the `p[]` syntax.

The point about objects with keys mapped to multiple values is important and will get dealt with next week.
